plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java

```diff
@@ -13,28 +13,111 @@
  */
 package org.apache.kudu.client;
 
+import io.trino.plugin.kudu.KuduClientSession;
+import io.trino.plugin.kudu.KuduClientWrapper;
 import io.trino.spi.TrinoException;
 
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkState;
 import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
 import static java.lang.String.format;
+import static org.apache.kudu.client.SessionConfiguration.FlushMode.MANUAL_FLUSH;
 
 /**
- * Operation.getChangeType() is package private
+ * This class is used to buffer operations and apply them in a batch and verify the batch was successfully applied.
+ * This gives:
+ * - performance of not flushing after every operation
+ * - correctness of not flushing in the background and hoping operations succeed
+ * The buffer is flushed when:
+ * - The KuduOperationApplier is closed
+ * - Or if the KuduOperationApplier reaches the bufferMaxOperations
+ * Note: Operation.getChangeType() is package private
  */
 public final class KuduOperationApplier
+        implements AutoCloseable
 {
-    private KuduOperationApplier()
+    private static final int bufferMaxOperations = 1000;
+
+    private int currentOperationsInBuffer;
+    private final KuduSession kuduSession;
+
+    private KuduOperationApplier(KuduSession kuduSession)
     {
+        kuduSession.setFlushMode(MANUAL_FLUSH);
+        kuduSession.setMutationBufferSpace(bufferMaxOperations);
+        this.kuduSession = kuduSession;
+        currentOperationsInBuffer = 0;
     }
 
-    public static OperationResponse applyOperationAndVerifySucceeded(KuduSession kuduSession, Operation operation)
+    public static KuduOperationApplier fromKuduClientWrapper(KuduClientWrapper kuduClientWrapper)
+    {
+        KuduSession session = kuduClientWrapper.newSession();
+        return new KuduOperationApplier(session);
+    }
+
+    public static KuduOperationApplier fromKuduClientSession(KuduClientSession kuduClientSession)
+    {
+        KuduSession session = kuduClientSession.newSession();
+        return new KuduOperationApplier(session);
+    }
+
+    public void applyOperationAsync(Operation operation)
```

Comment:
Can multiple threads call this concurrently?

If so what happens to consistency if one batch of operations fails and the other does not?

(might be a stupid question).

Reply:
KuduSession is not thread safe: https://kudu.apache.org/apidocs/org/apache/kudu/client/KuduSession.html

See the comment above, the assumption is this should never be called concurrently.

Reply:
Do mention this in the class level and this method's javadoc (if we don't end up adding @NotThreadSafe).